man/hpttree.1
.TH HPTTREE 1 "08 Jan 2000"
.SH NAME
hpttree \- area distribution tree builder for hpt
.SH SYNOPSIS
.B hpttree
[\-c config] [\-p] [areaname ...]

.SH DESCRIPTION
hpttree scan areas bases and prints distribution tree of areas specified
by arguments \fIareaname\fP.

.SS Options
.TP
.B \-c <config>
husky config file.
.TP
.B \-p
toggle pseudographics mode. Default for UNIXs is do not use pseudographic
character, and use that ones on other platforms.


.SH ENVIROMENT
.B FIDOCONFIG
.PP
path/filename of the husky configuration file

.SH FILES
.I /some/path/config
.B husky
configuration file
.PP
.I /MsgBaseDir/<areabase>
area base file or directory name

.SH SEE ALSO
.B hpt(1)
.PP
.B hpt
site:
.I http://husky.physcip.uni-stuttgart.de/hpt.html

.SH AUTORS
.B Serguei Revtov
2:5021/11.10@FIDOnet or 2:5021/19.1@FIDOnet

.B The Husky Team:

.I http://husky.physcip.uni-stuttgart.de/team.html
.TP
and some other peoples in the world.



HISTORY
HPT 1.4 stable
Changes between hpt-1.2-stable and hpt-1.4-stable

1. MAJOR feature improvements:
   - new parameter for bundlenamestyle: addrsCRC32 - for generate more unique
     bundle file names for every link pair (thanks to Igor Romanovsky);
   - Don't include Origin into areafix reply if areafixorigin is empty string;
   - reduces SEEN-BY (FSC-0093)
   - new areafix command: %rsb on/off - turns on/off reduced SEEN-BY mode.
   - @RESCANNED kludge on rescanned messages (FSC-0057)

2. CONFIG changes:
   - area definition options -manual & -mandatory are different:
     '-mandatory' prevent unsubscribe, '-manual' disable subscribe;
   - remove round quotes from origin & areafixorigin; for define empty origin
     set to "";
   - new feature: NewAreaRefuseFile - file with masks of areas which we don't
     want to autocreate;
   - new keyword: disablePID: Same as disableTID, but works only for messages
     being posted;
   - new keyword: AvailList (link section): specify variant for reply to %avail
     command.
   - ^AFLAGS NPD is now configurable via config (areafixReportsAttr)

3. IMPORTANT bugfixes:
   - deadareas.pl: Change dupebase files suffix: 'dpd' to 'dph' ('dpd' is
     obsolete);
   - do not scan messages with flag 'LOK' set;
   - If message is being scanned (i.e. has 'Loc' flag and doesn't have 'Snt'
     flag) and contains one or more @TID kludges all of them will be
     replaced by one kludge with info about hpt;
   - Posted message doesn't contain ^ATID anymore. It will be added only during
     scan;
   - Fix trap for messages with illegal date field value: begin of epoch or
     very big (<1.1.1970 and >n.n.2038)
   - Check netmailarea presence in config (before: trap)
   - ImportLog will not contain passthrough areas any more.
   - Fix CarbonCopy processing: add end of line after kludges
   - Fix linking when ^AMSGID or ^AREPLY are in MiXeD cAsE
   - keepTrsMail works now
   - If hpt is being run several times within the same second it could generate
     the same pkt name and archiver will overwrite existing pkt in bundle.
     Now before generating 1st pkt hpt make a delay in 1 second to prevent 
     this. Or set seqDir in config and no delay will be at all! :)
   - Changed zonegating algorythm (see docs)
   - Take from/to addresses from INTL instead of binary headers when packing 
     netmail from message base. If addresses from kludges not available - 
     addresses from headers will be taken.
   - Fixed bug with changing 1st line of included config file while
     autocreating area
   - Fix buggy linking when ^AMSGID contains 5d address
   - Send unsubscribe message when area going from idle state to kill
     (areafixQueueFile)
   - Send "-area" instead of "~area" when updating areafixQueueFile even if
     advancedAreaFix is on for uplink.
   - "carbonrule delete" may remove message from (common) netmail, not  only
     from specified area


HPT 1.2.3-release
(1.2-stable branch started 25.06.2002)
Changes between hpt-1.0-stable and hpt-1.2-stable

1. MAJOR feature improvements:
   - new areafix code; idle forward requests (track new echoes requests
     from uplinks);
   - 'hpt qupd' - check requested echoes queue,
   - 'hpt qrep' - report queue status (send requested echoes status to links)
   - areafix: report about already linked areas
   - areafix: send echoarea rules on subscribe (token sendRules)
   - areafix: new syntax of %list and %avail
   - toss speedup
   - win32 binary with dll (MS Visual C/C++ compiler)
   - Compile HPT with Perl support on MS VisualC & VisualC.NET
   - new perl hook tossbad(), called when put message to badarea
   - vars $fromname, $toname, $date, $subj for route() func in filter.pl
   - flags 'dir', 'imm', 'kfs', 'tfs', 'cfm' and 'npd' support
   - hpt post: uue-multipart posting
   - hptlink use hpt.log for logging
   - misc/hptlogstat.pl: new version
   - misc/pretty.pl: sort & format area definitions
   - misc/hpt_ro: New perl module for filter.pl: hpt read-only manager
     (written by Andrew Sagulin 2:6073/1).
   - added hpt link -j switch, link jam areas by msgid/reply crc -- speedup

2. CONFIG changes:
   - the 'pause' keyword syntax expanded: on, off, 1, 0, earea or farea
     possible ('pause earea': echoes paused, 'pause farea': fileareas paused)
   - change loglevels (see fidoconf & hpt documentation)
   - new keyword for link: NoRules
   - add sample configs to distribution

3. IMPORTANT bugfixes:
   - do not autoadd tearline & origin in netmail;
   - do not scan messages with flag 'LOK' set (mfc);
   - Clear trs, loc & k/s attributes after scan() perl hook;
   - translates origin & tirline to OEM-defined character set (win32);
   - HighWaterMark fix where scan jam area with deleted messages;
   - hptlink: ignore characters case and .0 at the end of node address
     in msgid/reply fields;
   - Fix remap (replace INTL kludge);
   - Fix trap on messages with illegal date field value: begin of epoch or
     very big (<1.1.1970 and >n.n.2038);
   - Check netmailarea presence in config (before: trap)
   - ImportLog will not contain passthrough areas any more.

$Id$


src/hpttree.c
/*
 *      hpttree - area tree builder for Highly Portable Tosser (hpt)
 *      by Serguei Revtov 2:5021/11.10 || 2:5021/19.1
 *
 * This file is part of HPT.
 *
 * HPT is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * HPT is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with HPT; see the file COPYING.  If not, write to the Free
 * Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *****************************************************************************
 * $Id$
 */

#include <ctype.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

#include <huskylib/compiler.h>
#include <huskylib/huskylib.h>

#ifdef HAS_IO_H
#include <io.h>
#endif

#ifdef HAS_UNISTD_H
#include <unistd.h>
#endif

#ifdef HAS_SHARE_H
#include <share.h>
#endif

#ifdef HAS_SYS_SYSEXITS_H
#include <sys/sysexits.h>
#endif
#ifdef HAS_SYSEXITS_H
#include <sysexits.h>
#endif

#include <smapi/msgapi.h>

#include <fidoconf/fidoconf.h>
#include <fidoconf/common.h>

#include <version.h>
#include "../cvsdate.h"

s_fidoconfig * cfg;
struct nodepath
{
    unsigned int zone, net, node;
    int          exportto;
    int          printed;
};

typedef struct nodepath s_nodepath;
s_nodepath * allNodes = NULL;
int nodeCount         = 0;
FILE * outlog;
char * versionStr  = NULL;
int * linksOnLevel = NULL;
int linksInArray   = 0;

#ifdef __UNIX__
int charsPG = 0;
#else
int charsPG = 1;
#endif

#define LLCORNER allCharsPG[charsPG][0]
#define VLINE allCharsPG[charsPG][1]
#define LTEE allCharsPG[charsPG][2]

char * allCharsPG[2] =
{
    "L|+", "\300\263\303"
};
int tperiod;
void printTree(int level, int nodeNum)
{
    int i;
    s_nodepath * cnode;
    int linech;

    cnode = &allNodes[nodeNum];

    if(linksInArray <= level)     /*  Allocate place for link counter in current level */
    {
        linksInArray = level + 1;
        linksOnLevel = srealloc(linksOnLevel, sizeof(int) * linksInArray);
    }

    linksOnLevel[level] = 0;

    for(i = 0; i < nodeCount; i++)
    {
        if((allNodes[i]).exportto == nodeNum)
        {
            linksOnLevel[level]++;
        }
    }

    if(level > 0)
    {
        linksOnLevel[level - 1]--;           /*  current node is already printed */
    }

    for(i = 0; i < level; i++)
    {
        linech = ' ';

        if(linksOnLevel[i])
        {
            linech = VLINE;
        }

        if(i == (level - 1))
        {
            linech = LTEE;

            if(linksOnLevel[level - 1] == 0)
            {
                linech = LLCORNER;
            }
        }

        printf("  %c ", linech);
    }
    printf("%u:%u/%u\n", cnode->zone, cnode->net, cnode->node);

    if(cnode->printed)
    {
        printf("WARNING: Loop at %u:%u/%u, escaping thread\n", cnode->zone, cnode->net,
               cnode->node);
        return;
    }

    cnode->printed = 1; /*  for checking for lost nodes */

    for(i = 0; i < nodeCount; i++)
    {
        if((allNodes[i]).exportto == nodeNum)
        {
            printTree(level + 1, i);
        }
    }
} /* printTree */

void buildAreaTree(s_area * area)
{
    HAREA harea;
    dword highMsg;
    HMSG hmsg;
    XMSG xmsg;
    unsigned int nmsg;
    char * text;
    dword textLength;
    s_nodepath node =
    {
        2, 0, 0, -1, 0
    };
    s_nodepath * cnode;
    int prevNode;
    char * start;
    char * endptr;
    unsigned long temp;
    int done;
    int root = -1;
    /* for time period */
    struct tm tmTime;
    time_t ttime, actualTime = time(NULL);

    fprintf(outlog, "Distribution tree of area %s", area->areaName);

    if(tperiod)
    {
        fprintf(outlog, " for last %d %s\n\n", tperiod, tperiod == 1 ? "day" : "days");
    }
    else
    {
        fprintf(outlog, "\n\n");
    }

    if((area->msgbType & MSGTYPE_PASSTHROUGH) == MSGTYPE_PASSTHROUGH)
    {
        fprintf(outlog, "PASSTHROUGH, ignoring\n");
        return;
    }

    if((area->useAka) != NULL)
    {
        node.zone = (area->useAka)->zone;
    }

/*   else
      node.zone = 2;
 */
    nodeCount    = linksInArray = 0;
    linksOnLevel = NULL;
    allNodes     = NULL;
    harea        = MsgOpenArea((byte *)area->fileName, MSGAREA_NORMAL, (word)area->msgbType);

    if(harea)
    {
        int i = -1;
        highMsg = MsgGetHighMsg(harea);

        if(highMsg < 1)
        {
            fprintf(outlog, "nothing to build (%ld messages)\n", (long)highMsg);
            MsgCloseArea(harea);
            return;
        }

        /* Pass 1: fill nodes array and buld links */
        for(nmsg = 1; nmsg <= highMsg; nmsg++)
        {
            hmsg = MsgOpenMsg(harea, MOPEN_READ, nmsg);

            if(hmsg)
            {
                /*
                 * NOTE: text would be destructed by strtoc() !!!
                 * Do not use it for anything else
                 */
                textLength = MsgGetTextLen(hmsg);
                text       = (char *)scalloc(textLength + 1, sizeof(char));

                if(text == NULL)
                {
                    return;
                }

                ttime = mktime(&tmTime);
                MsgReadMsg(hmsg, &xmsg, 0, textLength, (unsigned char *)text, 0, NULL);

                /* check time period */
                if(xmsg.attr & MSGLOCAL)
                {
                    DosDate_to_TmDate((SCOMBO *)&(xmsg.date_written), &tmTime);
                }
                else
                {
                    DosDate_to_TmDate((SCOMBO *)&(xmsg.date_arrived), &tmTime);
                }

                MsgCloseMsg(hmsg);

                /* check time period */
                if((tperiod) && (labs(actualTime - ttime) >= (tperiod * 24 * 60 * 60)))
                {
                    continue;
                }

                start    = text;
                prevNode = -1;
                done     = 0;

                /*  find beginning of path lines */
                do
                {
                    start = strstr(start, "\001PATH: ");

                    if(start == NULL)
                    {
                        done++;
                    }

                    if(!done)
                    {
                        start += 7; /*  jump over PATH: */

                        while(*start == ' ')
                        {
                            start++;        /*  find first word after PATH: */
                        }
                    }
                }
                while(!done && !isdigit((int)*start));

                if(!done)
                {
                    char * token;
                    token = strtok(start, " \r\t\376");

                    while(token != NULL && !done)
                    {
                        if(isdigit((int)*token))
                        {
                            int found;
                            /*  parse token */
                            temp = strtoul(token, &endptr, 10);

                            if((*endptr) == ':') /*  zone */
                            {
                                node.zone = temp;
                                endptr++;
                                temp = strtoul(endptr, &endptr, 10);
                            }

                            if((*endptr) == '/') /*  net */
                            {
                                node.net = temp;
                                endptr++;
                                temp = strtoul(endptr, &endptr, 10);
                            }

                            if(*endptr)
                            {
                                fprintf(outlog,
                                        "POINT or bad address in PATH: in message %u\n",
                                        nmsg);
                            }

                            /*  only node aka */
                            node.node = temp;

                            /*  find if there where that node in array */
                            for(found = 0, i = 0, cnode = allNodes; i < nodeCount && !found;
                                i++, cnode++)
                            {
                                if(cnode->node == node.node && cnode->net == node.net &&
                                   cnode->zone == node.zone)
                                {
                                    found++;
                                }
                            }

                            if(!found)
                            {
                                nodeCount++;
                                i        = nodeCount;
                                allNodes = (s_nodepath *)srealloc(allNodes,
                                                                  sizeof(s_nodepath) * nodeCount);
                                cnode  = &allNodes[i - 1];
                                *cnode = node;
                            }

                            i--; /*  was incremented in 'for' & in prev. 'if' */

                            if(prevNode >= 0 && prevNode != i)
                            {
                                (allNodes[prevNode]).exportto = i;
                            }

                            prevNode = i;
                        }
                        else
                        {
                            if(strchr(" \r\t\376",
                                      *token) == NULL && strncmp(token, "\001PATH:", 6) != 0)
                            {
                                done++;                                                              /*

                                                                                                        something's
                                                                                                        wrong
                                                                                                        */
                            }
                        }

                        token = strtok(NULL, " \r\t\376");
                    }

                    if(root < 0)
                    {
                        root = i;
                    }
                }

                nfree(text);
            }
        }
        MsgCloseArea(harea);

        /*  printing tree */
        if(nodeCount > 0)
        {
            printTree(0, root);
        }
        else
        {
            printf("Not distributed\n");
        }

        for(i = 0; i < nodeCount; i++)
        {
            if(!((allNodes[i]).printed))
            {
                fprintf(outlog, "Lost Node: %d:%d/%d\n", (allNodes[i]).zone, (allNodes[i]).net,
                        (allNodes[i]).node);
            }
        }
        nfree(allNodes);
        nfree(linksOnLevel);
        fprintf(outlog, "\nGenerated by %s\n", versionStr);
    }
    else
    {
        fprintf(outlog, "\nCould not open area %s\n", area->areaName);
    }
} /* buildAreaTree */

void usage(void)
{
    printf(
        "Usage: hpttree [options] [areaname ...]\n"
        "Options:\n"
        "\t  -c <config>\t- config file\n"
        "\t  -p - toggle pseudographics mode\n"
        "\t  -d <num>\t- for last <num> days\n");
}

int main(int argc, char ** argv)
{
    unsigned int i;
    int j;
    struct _minf m;
    char ** argareas = NULL;
    int nareas       = 0;
    int found;
    const char * config = NULL;

    outlog = stdout;
    setbuf(outlog, NULL);
    versionStr = GenVersionStr("hpttree", hpt_VER_MAJOR, hpt_VER_MINOR,
                               hpt_VER_PATCH, hpt_VER_BRANCH, cvs_date);
    printf("%s\n\n", versionStr);

    for(j = 1; j < argc; j++)
    {
        if(argv[j][0] == '-')
        {
            switch(argv[j][1])
            {
                case 'c':
                case 'C':
                    j++;
                    if(!argv[j])
                    {
                        usage();
                        exit(EX_USAGE);
                    }
                    config = argv[j];
                    break;

                case 'p': /* Toggle pseudographics */
                case 'P':
                    charsPG = (charsPG) ? 0 : 1;
                    break;

                case 'd': /* Last NUM days */
                case 'D':
                    j++;

                    if(!argv[j])
                    {
                        usage();
                        exit(EX_USAGE);
                    }

                    tperiod = atoi(argv[j]);

                    if(tperiod <= 0)
                    {
                        usage();
                        exit(EX_USAGE);
                    }

                    break;

                default:
                    usage();
                    exit(EX_USAGE);
            } /* switch */
        }
        else
        {
            /*  AreaName(s) specified by args */
            nareas++;
            argareas             = (char **)srealloc(argareas, nareas * sizeof(char *));
            argareas[nareas - 1] = argv[j];
        }
    }
    cfg = readConfig(config);

    if(!cfg)
    {
        fprintf(outlog, "Could not read fidoconfig\n");
        return 1;
    }

    m.req_version = 0;
    m.def_zone    = (UINT16)cfg->addr[0].zone;

    if(MsgOpenApi(&m) != 0)
    {
        fprintf(outlog, "MsgOpenApi Error.\n");
        exit(EX_SOFTWARE);
    }

    if(argareas)
    {
        /*  link only specified areas */
        for(j = 0; j < nareas; j++)
        {
            found = 0;

            /*  EchoAreas */
            for(i = 0; i < cfg->echoAreaCount && !found; i++)
            {
                if(stricmp(cfg->echoAreas[i].areaName, argareas[j]) == 0)
                {
                    buildAreaTree(&(cfg->echoAreas[i]));
                    found++;
                }
            }

            if(!found)
            {
                fprintf(outlog, "Couldn't find area \"%s\"\n", argareas[j]);
            }
        }
    }
    else
    {
        /*  EchoAreas */
        for(i = 0; i < cfg->echoAreaCount; i++)
        {
            buildAreaTree(&(cfg->echoAreas[i]));
        }
    }

    disposeConfig(cfg);
    return 0;
} /* main */
